contd: read environment variable from .stratup.toml if it exists and ensure /etc/resolv.conf

zaibon opened the pull request.

zaibon committed:
contd: read environment variable from .stratup.toml if it exists

zaibon committed:
sort list before comparing

muhamadazmy commented on pkg/container/container.go:
```diff
@@ -157,6 +162,19 @@ func (c *containerModule) Run(ns string, data pkg.Container) (id pkg.ContainerID
 		return id, ErrEmptyRootFS
 	}
 
+	startupFile, err := os.Open(filepath.Join(data.RootFS, ".startup.toml"))
```
Can this be moved to another method, it looks like a lot of code in the Run method . Something like `updateEnv(&data, path.Join(root, ".startup.toml"))`

muhamadazmy commented on pkg/container/container.go:
```diff
@@ -157,6 +162,19 @@ func (c *containerModule) Run(ns string, data pkg.Container) (id pkg.ContainerID
+
+		data.Env = mergeEnvs(envs, data.Env)
+	}
+
```
I see that u still do not use the `exec` from the `startup.toml` as an entrypoint for the container.
The procedure should be the following
- check if .startup.toml exists (opens with no error)
- load the `entry` point.
- if an `entrypoint` is already provided by the reservation, use it.
- if no `entrypoint` is provided, use the one from .startup.tom
- update the env

zaibon committed:
contd: parse startup.toml and update container data

fixes #319

we extract 3 things from statup.toml:
- entrypoint
- working dir
- enviroment variable

muhamadazmy commented on pkg/container/container.go:
```diff
@@ -298,3 +309,72 @@ func (c *containerModule) Delete(ns string, id pkg.ContainerID) error {
+	return service.Create(ctx, namespace, nil)
+}
+
+func setResolvConf(root string) error {
```
I don't get what this method is trying to do (i know it is trying to write the etc/resolv.conf with static content) but the check and the stats, i don't get. I mean why exactly didn't u just use `ioutil.WriteFile` method. ?

zaibon replied:
Cause I don't want to overwrite the content if there is any. So here I'm writing only of the file doesn't exist or if it's empty

muhamadazmy replied:
Does docker do that? because to me this sound weird. A container should always have a valid dns related to it's runtime network, not build time network. So this sounds a bit weird to me.

maxux replied:
Docker mount host resolv.conf to container resolv.conf using volume mount (AFAIK)

zaibon replied:
yes as far as I've seen. Docker does write a file on the host filesystem and mounts it inside the container when it starts.
Docker does the same for hosts hostname and hosts file.

I can also do it the same way if you prefer. It just that since we already have a copy on write layer with flist I though maybe this is easier.

But the containerd has indeed a set of methods to deal with such a things:
```go
// WithHostHostsFile bind-mounts the host's /etc/hosts into the container as readonly
func WithHostHostsFile(_ context.Context, _ Client, _ *containers.Container, s *Spec) error {
	s.Mounts = append(s.Mounts, specs.Mount{
		Destination: "/etc/hosts",
		Type:        "bind",
		Source:      "/etc/hosts",
		Options:     []string{"rbind", "ro"},
	})
	return nil
}

// WithHostResolvconf bind-mounts the host's /etc/resolv.conf into the container as readonly
func WithHostResolvconf(_ context.Context, _ Client, _ *containers.Container, s *Spec) error {
	s.Mounts = append(s.Mounts, specs.Mount{
		Destination: "/etc/resolv.conf",
		Type:        "bind",
		Source:      "/etc/resolv.conf",
		Options:     []string{"rbind", "ro"},
	})
	return nil
}

// WithHostLocaltime bind-mounts the host's /etc/localtime into the container as readonly
func WithHostLocaltime(_ context.Context, _ Client, _ *containers.Container, s *Spec) error {
	s.Mounts = append(s.Mounts, specs.Mount{
		Destination: "/etc/localtime",
		Type:        "bind",
		Source:      "/etc/localtime",
		Options:     []string{"rbind", "ro"},
	})
	return nil
}
```

muhamadazmy commented on pkg/container/container.go:
```diff
@@ -298,3 +309,72 @@ func (c *containerModule) Delete(ns string, id pkg.ContainerID) error {
+
+func applyStartup(data *pkg.Container, path string) error {
+	f, err := os.Open(path)
+	if err == nil {
```
instead of indenting all the method implementation, it would better if u do 
```go
if err != nil {
    return err
}
defer f.Close()
// everything else goes here.
```
no the method just look weird can confusing.

zaibon replied:
Well this is on purpose cause I really just want to do something when the file exists.
If it doesn't the this function does nothing. I though having it in the `err == nil` block would make this more obvious.

I can change it to a `os.IsNotExists` check if you prefer.

muhamadazmy replied:
Oh, i see. in that case i think it's fine.

muhamadazmy commented on pkg/container/startup.go:
```diff
@@ -0,0 +1,68 @@
+// all the key from a will endup in b
+// if a key is present in both, key from a are kept
+func mergeEnvs(a, b []string) []string {
+	ma := make(map[string]string, len(a))
```
Why this has to be 2 dicts instead of one, the merge can done in the 2 loops directly but now u need 3 loops instead. Also did u test what will happen if u have `["KEY=V1", "KEY=V2"]` passed directly to process Env, because i think libc will just do the dedubing for u. this can be easily tested with one exec.Command line of `env`

zaibon replied:
> Why this has to be 2 dicts instead of one, the merge can done in the 2 loops directly but now u need 3 loops instead. 

Right... :-)

>  Also did u test what will happen if u have ["KEY=V1", "KEY=V2"] passed directly to process Env, because i think libc will just do the dedubing for u.

I don't want to rely on a side effect when I can be sure to do it myself here.

muhamadazmy reviewed (changes requested):
More few comments
